From: "Jan Alexander Steffens (heftig)" <jan.steffens@gmail.com>
Date: Tue, 27 Oct 2020 19:50:30 +0000
Subject: screen: Manually handle env inheritance

Don't let VTE handle merging the environment; build the merged block
ourselves. This avoids confusion when we reuse the exec data.

Bug: https://gitlab.gnome.org/GNOME/gnome-terminal/-/issues/303
Bug-Debian: https://bugs.debian.org/973918
Applied-upstream: no
---
 src/terminal-screen.c | 26 +++++++++++++-------------
 1 file changed, 13 insertions(+), 13 deletions(-)

diff --git a/src/terminal-screen.c b/src/terminal-screen.c
index 0c4ee50..d44315c 100644
--- a/src/terminal-screen.c
+++ b/src/terminal-screen.c
@@ -945,9 +945,7 @@ terminal_screen_exec (TerminalScreen *screen,
                                                                   &shell);
 
   gboolean preserve_cwd = FALSE;
-  GSpawnFlags spawn_flags = G_SPAWN_SEARCH_PATH_FROM_ENVP;
-  if (initial_envv)
-    spawn_flags |= VTE_SPAWN_NO_PARENT_ENVV;
+  GSpawnFlags spawn_flags = G_SPAWN_SEARCH_PATH_FROM_ENVP | VTE_SPAWN_NO_PARENT_ENVV;
   gs_strfreev char **exec_argv = NULL;
   if (!terminal_screen_get_child_command (screen,
                                           argv,
@@ -1434,6 +1432,7 @@ terminal_screen_get_child_environment (TerminalScreen *screen,
 {
   TerminalApp *app = terminal_app_get ();
   char **env;
+  gs_strfreev **current_environ = NULL;
   char *e, *v;
   GHashTable *env_table;
   GHashTableIter iter;
@@ -1442,17 +1441,18 @@ terminal_screen_get_child_environment (TerminalScreen *screen,
 
   env_table = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_free);
 
-  env = initial_envv;
-  if (env)
+  if (initial_envv)
+    env = initial_envv;
+  else
+    env = current_environ = g_get_environ ();
+
+  for (i = 0; env[i]; ++i)
     {
-      for (i = 0; env[i]; ++i)
-        {
-          v = strchr (env[i], '=');
-          if (v)
-             g_hash_table_replace (env_table, g_strndup (env[i], v - env[i]), g_strdup (v + 1));
-           else
-             g_hash_table_replace (env_table, g_strdup (env[i]), NULL);
-        }
+      v = strchr (env[i], '=');
+      if (v)
+          g_hash_table_replace (env_table, g_strndup (env[i], v - env[i]), g_strdup (v + 1));
+        else
+          g_hash_table_replace (env_table, g_strdup (env[i]), NULL);
     }
 
   g_hash_table_remove (env_table, "COLUMNS");
